tensorflow_federated/python/core/impl/executors/federating_executor.py

```diff
@@ -453,6 +451,221 @@ def validate_executor_placements(cls, executor_placements):
+    return await self._place(arg, placement_literals.CLIENTS)
+
+  async def federated_eval_at_server(self, arg):
+    return await self._eval(arg, placement_literals.SERVER, True)
+
+  async def federated_eval_at_clients(self, arg):
+    return await self._eval(arg, placement_literals.CLIENTS, False)
+
+  async def federated_apply(self, arg):
+    return await self._map(arg)
+
+  async def federated_map(self, arg):
+    return await self._map(arg, all_equal=False)
+
+  async def federated_map_all_equal(self, arg):
+    return await self._map(arg, all_equal=True)
+
+  async def federated_broadcast(self, arg):
+    py_typecheck.check_type(arg.type_signature, computation_types.FederatedType)
+    py_typecheck.check_type(arg.internal_representation, list)
+    if not arg.type_signature.all_equal:
+      raise ValueError('Cannot broadcast a non all_equal value.')
+    if len(arg.internal_representation) != 1:
+      raise ValueError(
+          'Cannot broadcast a with a non-singleton representation.')
+    val = await arg.internal_representation[0].compute()
+    return FederatingExecutorValue(
+        await asyncio.gather(*[
+            c.create_value(val, arg.type_signature.member)
+            for c in self._get_child_executors(placement_literals.CLIENTS)
+        ]), type_factory.at_clients(arg.type_signature.member, all_equal=True))
+
+  async def federated_zip_at_server(self, arg):
+    return await self._zip(arg, placement_literals.SERVER, all_equal=True)
+
+  async def federated_zip_at_clients(self, arg):
+    return await self._zip(arg, placement_literals.CLIENTS, all_equal=False)
+
+  async def federated_reduce(self, arg):
+    self._check_arg_is_anonymous_tuple(arg)
+    if len(arg.internal_representation) != 3:
+      raise ValueError(
+          'Expected 3 elements in the `federated_reduce()` argument tuple, '
+          'found {}.'.format(len(arg.internal_representation)))
+
+    val_type = arg.type_signature[0]
+    py_typecheck.check_type(val_type, computation_types.FederatedType)
+    item_type = val_type.member
+    zero_type = arg.type_signature[1]
+    op_type = arg.type_signature[2]
+    type_utils.check_equivalent_types(
+        op_type, type_factory.reduction_op(zero_type, item_type))
+
+    val = arg.internal_representation[0]
+    py_typecheck.check_type(val, list)
+    aggr = self._get_child_executors(placement_literals.AGGREGATOR, index=0)
+    server = self._get_child_executors(placement_literals.SERVER, index=0)
+
+    async def _move(v, target_executor):
+      return await target_executor.create_value(await v.compute(), item_type)
+
+    aggregands = await asyncio.gather(*[_move(v, aggr) for v in val])
+
+    zero = await aggr.create_value(
+        await (await self.federating_executor.create_selection(
+            arg, index=1)).compute(),
+        zero_type)
+    op = await aggr.create_value(arg.internal_representation[2], op_type)
+
+    result = zero
+    for item in aggregands:
```

Comment:
another place when logic is mixed between driver and aggregator; think we should at least have a discussion about this

Reply:
Originally I had an explicit move op from the aggregator to server at the end before embedding in the executor, but I think you mentioned it was redundant. Happy to have a conversation but I actually came around to this view (although maybe that's not what you're referring to)

Reply:
On second glance I don't see what you're referring to as being mixed between the two

Reply:
what I mean is that the aggregator just sees a bunch of (binary) additions without context, and its only the driver that understands that we're during a reduction. so if the driver wanted to eg add the same values more than once the aggregator would have no clue (not the best example)

Reply:
seems like that's unavoidable though? the federated intrinsic for reduce is defined in such a way that we're forced into building it from a binary operator. I'm not sure I see a way of shipping e.g. the whole reduce option as one Call object to the aggregator. happy to investigate more if we have any good leads

Reply:
I think all of the reduce logic could be expressed as a `tf_computation`, but also happy with letting this one simmer!